src/dfx/src/lib/builders/mod.rs

```diff
@@ -324,27 +324,49 @@ fn ensure_trailing_newline(s: String) -> String {
     }
 }
 
-pub fn run_command(args: Vec<String>, vars: &[Env<'_>], cwd: &Path) -> DfxResult<()> {
-    let (command_name, arguments) = args.split_first().unwrap();
-    let canonicalized = dfx_core::fs::canonicalize(&cwd.join(command_name))
-        .or_else(|_| which::which(command_name))
-        .map_err(|_| anyhow!("Cannot find command or file {command_name}"))?;
-    let mut cmd = Command::new(canonicalized);
-
-    cmd.args(arguments)
-        .current_dir(cwd)
-        .stdout(Stdio::inherit())
-        .stderr(Stdio::inherit());
-
+/// Run a command and return its output.
+/// If the io_inherit is true, the return bytes will always be empty.
+pub fn run_command(
+    command: &str,
+    vars: &[Env<'_>],
+    cwd: &Path,
+    io_inherit: bool,
+) -> DfxResult<Vec<u8>> {
+    // No commands, noop.
+    if command.is_empty() {
+        return Ok(vec![]);
+    }
+    let (first, rest) = match command.split_once(' ') {
+        Some((first, rest)) => (first, rest),
+        None => (command, ""),
+    };
+
+    let canonicalized = dfx_core::fs::canonicalize(&cwd.join(first))
+        .or_else(|_| which::which(first))
+        .map_err(|_| anyhow!("Cannot find command or file {first}"))?;
+    let canonicalized_command = [
+        canonicalized.to_string_lossy().to_string(),
```

Comment:
Can we avoid passing this path through to_string_lossy()?  The result of `which()` in particular doesn't necessarily have to pass through losslessly. I think `cmd.arg(canonicalized).arg(rest)` will achieve this.